File: applications/backend/core/src/main/kotlin/com/forsetijudge/core/port/driven/broadcast/room/pprivate/JudgePrivateBroadcastRoom.kt
package com.forsetijudge.core.port.driven.broadcast.room.pprivate

import com.forsetijudge.core.domain.entity.Ticket
import com.forsetijudge.core.port.driven.broadcast.BroadcastEvent
import com.forsetijudge.core.port.dto.response.ticket.toResponseBodyDTO
import java.util.UUID

class JudgePrivateBroadcastRoom(
    contestId: UUID,
    memberId: UUID,
) {
    companion object {
        val pattern = Regex("/contests/(?<contestId>[a-f0-9\\-]+)/members/(?<memberId>[a-f0-9\\-]+)/private/judge")
    }

    private val name = "/contests/$contestId/members/$memberId/private/judge"

    fun buildTicketUpdatedEvent(ticket: Ticket<*>) =
        BroadcastEvent(
            room = name,
            name = "TICKET_UPDATED",
            data = ticket.toResponseBodyDTO(),
        )
}

File: applications/webapp/src/core/port/driven/broadcast/room/private/JudgePrivateBroadcastRoom.ts
import { BroadcastRoom } from "@/core/port/driven/broadcast/BroadcastRoom";
import { TicketResponseDTO } from "@/core/port/dto/response/ticket/TicketResponseDTO";

export type JudgePrivateBroadcastRoomCallbacks = {
  TICKET_UPDATED: (ticket: TicketResponseDTO) => void;
};

export class JudgePrivateBroadcastRoom extends BroadcastRoom<JudgePrivateBroadcastRoomCallbacks> {
  constructor(
    contestId: string,
    memberId: string,
    public callbacks: JudgePrivateBroadcastRoomCallbacks,
  ) {
    super(
      `/contests/${contestId}/members/${memberId}/private/judge`,
      callbacks,
    );
  }
}
